icu4c/source/common/ulocimp.h

```diff
@@ -307,4 +307,62 @@ U_CAPI const char* const* ulocimp_getKnownCanonicalizedLocaleForTest(int32_t* le
 // Return true if the value is already canonicalized.
 U_CAPI bool ulocimp_isCanonicalizedLocaleForTest(const char* localeName);
 
+/**
+ * A utility class for handling locale IDs that may be longer than ULOC_FULLNAME_CAPACITY.
+ * This encompasses all of the logic to allocate a temporary locale ID buffer on the stack,
+ * and then, if it's not big enough, reallocate it on the heap and try again.
+ *
+ * You use it like this:
+ * UErrorCode err = U_ZERO_ERROR;
+ *
+ * PreflightingLocaleIDBuffer tempBuffer;
+ * do {
+ *     *(tempBuffer.getCapacityPtr()) = uloc_doSomething(localeID, tempBuffer.getBuffer(), tempBuffer.getCapacity(), &err);
+ * } while (tempBuffer.needToTryAgain(&err));
+ * if (U_SUCCESS(err)) {
+ *     uloc_doSomethingWithTheResult(tempBuffer.getBuffer());
+ * }
+ */
+class PreflightingLocaleIDBuffer {
+private:
+    char stackBuffer[ULOC_FULLNAME_CAPACITY];
+    char* heapBuffer = nullptr;
+    int32_t capacity = ULOC_FULLNAME_CAPACITY;
+    
+public:
+    PreflightingLocaleIDBuffer() {}
+    
+    ~PreflightingLocaleIDBuffer() { delete [] heapBuffer; }
```

Comment:
Hm, I think this needs to also derive from the `UMemory` class in order to override the `new`/`delete` operators.
(Or you could use `uprv_malloc`/`uprv_free` directly.)

Reply:
Why do I need to override the new/delete operators?  What am I missing here?

Reply:
The ICU library code allows for overriding/changing the memory allocation functions, allowing the library code to possibly use a different heap than the application (or whatever is calling ICU).

There's a macro for controlling this called `U_OVERRIDE_CXX_ALLOCATION`, which has some comments about it:
https://github.com/unicode-org/icu/blob/2dce62892b6b00df02195a35ca09ebd8e082e80f/icu4c/source/common/unicode/uconfig.h#L114-L123

Also from the ICU user guide docs, there is this part here:
(From https://unicode-org.github.io/icu/userguide/dev/codingguidelines.html#c-class-layout-and-boilerplate )
> In general, all instantiable classes (i.e., all classes except for interface/mixin classes and ones with only static member functions) inherit the UMemory base class. UMemory provides new/delete operators, which allows to keep the ICU heap separate from the application heap, or to customize ICU’s memory allocation consistently.

Reply:
(sigh)  Okay, I changed it to use `uprv_malloc()` and `uprv_free()`.  I also added code to prevent heap allocation of the new class.  I'm hoping one or the other of these changes will also take care of the couple of build failures I'm seeing right now.

Reply:
Thanks! :)